### ISSUE TITLE
interactivate.py can not generate translation result

### ISSUE THREAD

**Issue by zll0000** [pytorch/fairseq#412]
The commond as follows:
python interactive.py data-bin/wmt14_en_de --path checkpoints/transformer/checkpoint_best.pt --beam 5 --remove-bpe --source-lang en --target-lang de 
however the interactivate.py can not generate translation result I want. The result as follows:
Namespace(beam=5, buffer_size=1, cpu=False, data=['/dnn4_added/zhangxiaolei/back_translation/fair-temp/fairseq//data-bin/wmt14_en_de'], diverse_beam_groups=1, diverse_beam_strength=0.5, fp16=False, fp16_init_scale=128, fp16_scale_window=None, gen_subset='test', left_pad_source='True', left_pad_target='False', lenpen=1, log_format=None, log_interval=1000, max_len_a=0, max_len_b=200, max_sentences=1, max_source_positions=1024, max_target_positions=1024, max_tokens=None, min_len=1, model_overrides='{}', nbest=1, no_beamable_mm=False, no_early_stop=False, no_progress_bar=False, num_shards=1, path='/dnn4_added/zhangxiaolei/back_translation/fair-temp/fairseq//checkpoints/transformer/checkpoint_best.pt', prefix_size=0, print_alignment=False, quiet=False, raw_text=False, remove_bpe='@@ ', replace_unk=None, sampling=False, sampling_temperature=1, sampling_topk=-1, score_reference=False, seed=1, shard_id=0, skip_invalid_size_inputs_valid_test=False, source_lang='en', target_lang='de', task='translation', unkpen=0, unnormalized=False, upsample_primary=1)
| [en] dictionary: 40360 types
| [de] dictionary: 42720 types
| loading model(s) from /dnn4_added/zhangxiaolei/back_translation/fair-temp/fairseq//checkpoints/transformer/checkpoint_best.pt
| Type the input sentence and press return:
Two sets of lights so close to one another : intentional or just a silly error ?
O       Two sets of lights so close to one another : intentional or just a silly error ?
H       -3.237684726715088      , , , , ,
P       -3.1607 -3.1607 -3.1607 -3.1607 -3.1607 -3.6224
Yesterday , Gutacht &apos;s Mayor gave a clear answer to this question .
O       Yesterday , Gutacht &apos;s Mayor gave a clear answer to this question .
H       -3.237684726715088      , , , , ,
P       -3.1607 -3.1607 -3.1607 -3.1607 -3.1607 -3.6224

The questions:
1.	I use the interactivate.py to translate the sequence I input , however only H line and P line are printed out ,where is the result 

I uses the generate.py without –reference-score :
python generate.py $TEXT/data-bin/wmt14_en_de --path $TEXT/checkpoints/transformer/checkpoint_best.pt --beam 5 --remove-bpe --source-lang en --target-lang de --gen-subset test 
output as follows:

S-1910  Just five minutes before kick @-@ off , the Capo climbed onto his podium - and the tension that had been almost tangible up until that point , lifted with a large shout , as Dickel asked for a shout from each section of fans as usual .
T-1910  Erst fünf Minuten vor dem Anpfiff erklomm der Capo sein Podest - und die bis dahin geradezu mit Händen greifbare Anspannung entlud sich in einem lauten Schrei , als Dickel wie gewohnt die Stimmung auf den Rängen im Stadion abfragte .
H-1910  -3.2376880645751953     , , , , ,
P-1910  -3.1608 -3.1608 -3.1608 -3.1608 -3.1608 -3.6224
| Translated 3003 sentences (18018 tokens) in 6.4s (469.52 sentences/s, 2817.09 tokens/s)                                                                                                           
| Generate test with beam=5: BLEU4 = 0.00, 24.8/0.0/0.0/0.0 (BP=0.037, ratio=0.233, syslen=15015, reflen=64506)

Question:
2.	The T line is from the test.de file ,which come reference file not the translation result. How to set the command can I get my models translation result

**Comment by huihuifan** [comment]
I will respond here instead of in your personal email, so if people have similar issues in the future, they can search and find my response. 

From the output you pasted, there is no problem with interactive.py. Since interactive.py is interactive, there is no "target" line. Otherwise, O represents your input, H the hypothesis, and P the token level log probabilities. This corresponds to the output you see from generate.py.

The *issue* is that your model does not seem to be well trained. As you can see from the output of both interactive.py and generate.py, the Hypothesis line is just commas. This indicates some sort of bad training. 

I would look into it in the following steps:
1- if you download one of our pretrained models and run the provided generation commands, does that work? 
2- please look at the training log of your model and make sure the loss is decreasing, etc. 

thanks.

**Comment by myleott** [comment]
Also if you trained your model with BPE, then you need to apply BPE to your input before passing it to interactive.py.

**Comment by gvskalyan** [comment]
Hi I am using a pretrained model English (en) to French (fr) based on Transformer in paper Ott et al.,2018 . Model is loaded through interactive.py the sentence to be given is said to be raw input. 
Is it Moses tokenized (normalize punctuation , clean corpus level clean the practical sentence, remove non-printing char , tokenizer ) and then BPE applied OR is  it just BPE applied and when I get hypothesis I need to do the reverse like remove bpe (but is already given as command line argumrnt and hence I suppose explicitly is not needed) and detokenize using moses .
suppose I give ::
O hi how are you
H	-0.8601110577583313	hi comment êtes @-@ vous
P	-0.9665 -2.3938 -1.1932 -0.1428 -0.0566 -0.4077

Where Am I going Wrong?

**Comment by gvskalyan** [comment]
Also I seem to be getting this when i use Interactive / generate on transformer pre-trained model 
--share-all-embeddings requires a joined dictionary
Please specify args when we use different architecture

**Comment by huihuifan** [comment]
preprocessing is detailed here: https://github.com/pytorch/fairseq/tree/master/examples/translation

please see the section on english-french.

You do not need to run the fairseq_incremental_decoder.py script yourself. The fconv model inherits from the incremental decoder class and decoding is already incremental if you run generate.py or interactive.py.

**Comment by gvskalyan** [comment]
Suppose I have run generate.py which uses binarised test , this is an example where H is the French output but it has these "&quot" and "&apos" which were due to pre processing can u please provide steps / doc  to post process So could get the French output.

S-100	The detractors are accusing the government of trying to save money by not allowing injured soldiers - who do not meet the army &apos;s rule of &quot; universality of service &quot; , which requires that personnel be able to carry out a series of varying tasks - to reach the ten @-@ year period of admissibility required for retirement benefits .
T-100	Les détracteurs accusent le gouvernement de vouloir économiser de l&apos; argent en ne permettant pas aux militaires blessés - qui ne répondent pas à la règle de &quot; l&apos; universalité du service &quot; de l&apos; armée exigeant que le personnel soit en mesure d&apos; effectuer une série de tâches variées - d&apos; atteindre la période d&apos; admissibilité de 10 ans requise pour les prestations de retraite .
H-100	-0.34499597549438477	Les détracteurs accusent le gouvernement d&apos; essayer de faire des économies en ne permettant pas aux soldats blessés - qui ne satisfont pas à la règle de l &apos; &quot; universalité du service &quot; de l&apos; armée , qui exige que le personnel soit en mesure d&apos; accomplir une série de tâches diverses - d&apos; atteindre la période d&apos; admissibilité de dix ans requise pour les prestations de retraite .
P-100	-0.4185 -0.0852 -0.0213 -0.0697 -0.1603 -0.0985 -0.1414 -0.1229 -1.0708 -0.3216 -0.9667 -0.4596 -0.1807 -0.1082 -0.1544 -0.8407 -0.2078 -0.1132 -0.2307 -0.1589 -0.0950 -0.4365 -0.1907 -0.1206 -1.3795 -0.1143 -0.3336 -0.1184 -0.1744 -0.6213 -0.9123 -0.7192 -0.1105 -0.0749 -0.2677 -0.0928 -0.1104 -0.5799 -0.1065 -0.2027 -0.4883 -0.7247 -0.3966 -0.1753 -0.4482 -0.1374 -0.8914 -0.7232 -0.0869 -0.1883 -0.6995 -0.5085 -0.1429 -0.1203 -0.1092 -1.0399 -0.0997 -0.3253 -0.7230 -0.2713 -0.1668 -0.4856 -0.0414 -0.1450 -0.2331 -0.0990 -0.8200 -0.2159 -1.7307 -0.1966 -0.1207 -0.0675 -0.1096 -0.1053

Also the docs here https://fairseq.readthedocs.io/en/latest/getting_started.html 
when used the same example and model (Pre-trained) in interactive.py, it gave Different outputs. Please clarify..
Input :: Why is it rare to discover new marine mam@@ mal species ?
output ::  Pourquoi est @-@ il rare de découvrir de nouvelles espèces de mammifères marins ?

SUpposed output (from docs ) :: Pourquoi est-il rare de découvrir de nouvelles espèces de mammifères marins ?

**Comment by gvskalyan** [comment]
But the type of preprocessing for English to French  based on scaling neural translation model is not given
And please state is their an easy way to state architecture as transformer base without specifying base args expilcitly

**Comment by gvskalyan** [comment]
can you please provide pre-trained models of scaling neural translation on fp16 @myleott

**Comment by myleott** [comment]
@gvskalyan, pretrained models are available here: https://github.com/pytorch/fairseq#translation. Those models were trained with FP16.

To preprocess the en-fr data, please follow the instructions here, adding `--joined-dictionary`: https://github.com/pytorch/fairseq/tree/master/examples/translation#prepare-wmt14en2frsh

The corresponding training command for 128 GPUs:
```
python train.py data-bin/wmt14_en_fr_joined \
  --arch transformer_vaswani_wmt_en_fr_big --share-all-embeddings \
  --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
  --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 \
  --lr 0.0007 --min-lr 1e-09 \
  --dropout 0.1 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
  --max-tokens 5120 \
  --fp16 --max-update 80000
```

If using fewer GPUs or smaller `--max-tokens`, then please set the `--update-freq` accordingly.

**Comment by myleott** [comment]
Re: Transformer base, you can use `transformer_wmt_en_de` and for Transformer big use `transformer_vaswani_wmt_en_de_big`. You can see all the architecture definitions here: https://github.com/pytorch/fairseq/blob/master/fairseq/models/transformer.py#L835-L915.

While these architectures work fairly well for many datasets, they are not universal. For smaller datasets you'll often need to use a smaller model to avoid overfitting. Similarly you may need to tune dropout and learning rate.

**Comment by gvskalyan** [comment]
How good is the export to caffe2 via ONNX to a fairseq 32/16 bit trained model on a CPU device using https://github.com/pytorch/translate
Is the inference a lot faster compared to pytorch models @myleott

**Comment by huihuifan** [comment]
@zll0000 there is some issue here. Preprocessing with --only-source is for language modeling. If you have a source sequence without reference, you can use interactive decoding. But the specific error you see here is because there is no directory called data-bin with your dict.jp.txt file in it...

**Comment by zll0000** [comment]
@huihuifan  thank you  !!I see

**Comment by masonreznov** [comment]
Hello @huihuifan @myleott,
I guess the hypothesis are generated according to the line length thus changing the original line number order. I want to know if there is any option to generate the hypothesis in the original line sequence, i.e the output should be in the folowing sequence
```
S-1
T-1
H-1
S-2
T-2
H-2 ... and so on.
```

**Comment by tamarakatic** [comment]
@masonreznov Hi, it is probably too late for the answer but it can help you in the future:

OUTPUT = "result.txt"
grep ^H $OUTPUT | sed 's/^H-//g' | sort -k1,1n | cut -f 3 | sed 's/ ##//g' > hyp.txt
grep ^T $OUTPUT | sed 's/^T-//g' | sort -k1,1n | cut -f 2 | sed 's/ ##//g' > ref.txt
